### ISSUE TITLE
Feature Proposal: Connect to Product Database

### ISSUE THREAD

**Issue by tderflinger** [reciperium/recipe-lang#12]
Hi there,

I think this recipe language is a great idea. When you specify an ingredient product, say an apple, it would be great to link it to some kind of database. The entry in the database for apple could then link to other sources like wikipedia entry and sources to buy the product.
Once you have linked the products in the recipe to a product database you can also much easier translate the recipe to other languages.

By the way, I have started what will become an open source online database of spices: https://spicenice.pages.dev/spices/all/

Maybe we can link the SpiceNice database with your recipe language to link the spice ingredients?

What do you think?

Best,
-Thomas

**Comment by woile** [comment]
Hi @tderflinger thanks for the compliments and the interest :) 

recipe-lang is composed of a parser (and a cli, but unrelated). The parser is able to read the language and output "rust" code. This rust code, can either be used to search for additional information or it can be given to another service (most probably as json) to add extra data. So I don't think "enriching" the recipe should happen here.

I've also built the [recipe-parser-wasm](https://github.com/reciperium/recipe-parser-wasm), which can be used in javascript to parse a recipe (which is lacking some docs, but I can add them if needed). With this library, you can parse recipe-lang, and get the javascript object, which at this point can be used to "enrich" the recipe

There's also a [react-recipe](https://github.com/reciperium/react-recipe) that can be used. Maybe I could add a plugin system to enrich the different types of items.

So I'm not entirely sure what to do here.

Some options I can think of:

- Building a prototype, similar to play.reciperium.com to test how this would behave, The playground is open source here: https://github.com/reciperium/playground
- Building a library that, given an ingredient, it would build the url to spicenice. I think approaching it from this side, it could help you design the api

Could you expand a bit more as of what do you have in mind?
Thanks!

**Comment by tderflinger** [comment]
![Architecture-FoodDB](https://github.com/user-attachments/assets/2c9e88ea-bc0e-4653-b0fb-5fea233e291e)

@woile Thanks for your nice message. I have added an architecture diagram of how I think about it. The basic idea is to create an open source food database that has information on food ingredients. The database itself has information about the ingredient but it also links to the food ingredient entry in Wikipedia and Wikidata and possibly other sources.

I have not found any existing open source food ingredient database out there. Do you know of any? So I think it would be valuable to build one. Yes, that takes time, but I think is worth it.

So once the food ingredient database exists, recipe-lang could query it. For example, I add "Tahini" or "lemon" to the recipe, recipe-lang queries the food database and gets the ID and a link to more information about the item.
When you generate the HTML of the recipe you could include links of each of the ingredient to the food database. When users click the link, they can read more about the ingredient (like a lemon or Tahini).

What do you think of the general idea? Is it worth it?

Best,
-Thomas

**Comment by woile** [comment]
I definitely think there should be some kind of solution for this, I know of the following:

- https://world.openfoodfacts.org/
- https://www.rivm.nl/en/dutch-food-composition-database
- https://fdc.nal.usda.gov/

And, again, one thing is reciperium.com and the other is recipe-lang (which is the parser).
Looking at the diagram, I don't want recipe-lang to do any queries, because it's out of the scope of the parser.
But again, you can build on top of recipe-lang and add all kinds of functionality, like hitting a database or an API, and I'd be glad to discuss or build some demos 👍🏻 

Eventually, when there's something mature in the ecosystem, rather than linking, I'd prefer to be able to calculate information about a recipe itself like:

- macros
- CO2 footprint
- analysis like: high in sodium, high in sugars, etc.

And for this I'd be very interested in having an open database!

**Comment by tderflinger** [comment]
@woile, yes, I see. I think other possibilities of a food database would be providing taste profiles, like how sour or sweet that ingredient is. There are more things like for example calculating calories.

So from my perspective, since I have already started with SpiceNice, I think I will first build out SpiceNice as a proof-of-concept. It consists only of a small part of all food ingredients (spices), but I think it is a good starting point.
I will let you know once I have added more features to SpiceNice. Would be great then, to explore some integration points!

Let's keep in contact! :)
